### ISSUE TITLE
FXGL modules to include version information

### ISSUE THREAD

**Issue by sormuras** [AlmasB/FXGL#1091]
Listing external modules of project https://github.com/sormuras/bach-fxgl reads like:
```
  com.almasb.fxgl.all
  com.almasb.fxgl.core
  com.almasb.fxgl.entity
  com.almasb.fxgl.gameplay
  com.almasb.fxgl.io
  com.almasb.fxgl.scene
  com.fasterxml.jackson.annotation@2.12.3
  com.fasterxml.jackson.core@2.12.3
  com.fasterxml.jackson.databind@2.12.3
  com.gluonhq.attach.audio@4.0.11
  com.gluonhq.attach.lifecycle@4.0.11
  com.gluonhq.attach.storage@4.0.11
  com.gluonhq.attach.util@4.0.11
  javafx.base
  javafx.controls
  javafx.fxml
  javafx.graphics
  javafx.media
  kotlin.stdlib
```
Only Jackson and Attach modules provide compiled version information. Let's compile the FXGL version into their module as well. Here's a similar feature request for JavaFX: https://bugs.openjdk.java.net/browse/JDK-8267472 (Perhaps, I'll raise such a feature request for Kotlin as well...)

**Comment by sormuras** [comment]
Adding the following lines to the root POM file:

```xml
                        <compilerArgs>
                            <arg>--module-version</arg>
                            <arg>123</arg>
                        </compilerArgs>
```

here:

https://github.com/AlmasB/FXGL/blob/842b079b2850b6b5444db8d983b88d9f32e263a6/pom.xml#L138-L145

...works as expected, for example:

```
jar --describe-module --file .../fxgl-scene/target/fxgl-scene-dev-SNAPSHOT.jar
com.almasb.fxgl.scene@123 jar:file:///.../FXGL/fxgl-scene/target/fxgl-scene-dev-SNAPSHOT.jar!/module-info.class
exports com.almasb.fxgl.notification
exports com.almasb.fxgl.notification.view
exports com.almasb.fxgl.scene
exports com.almasb.fxgl.ui
requires com.almasb.fxgl.core
requires java.base mandated
requires javafx.controls
qualified exports com.almasb.fxgl.notification.impl to com.almasb.fxgl.all
qualified opens com.almasb.fxgl.notification.impl to com.almasb.fxgl.core
qualified opens com.almasb.fxgl.scene to com.almasb.fxgl.core
qualified opens com.almasb.fxgl.ui to com.almasb.fxgl.core
```
...but only if the version (here hard-coded to the value `123`) is a valid Java module version string. `dev-SNAPSHOT` is not valid, as it doesn't start with digit/number: https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/lang/module/ModuleDescriptor.Version.html

Thus, using:

```xml
                        <compilerArgs>
                            <arg>--module-version</arg>
                            <arg>${project.version}</arg>
                        </compilerArgs>
```
fails for the default case:

`error: bad value for --module-version option: 'dev-SNAPSHOT'`

**Comment by sormuras** [comment]
Using `11+dev-SNAPSHOT` as the default version should do the trick.

**Comment by sormuras** [comment]
Would result in `com.almasb.fxgl.scene@11+dev-SNAPSHOT` and similar for the default case.

And `com.almasb.fxgl.scene@11.18` for the next "dot release".

Also, the next "major release" version string would read for example: `17+dev-SNAPSHOT` / `17` / `17.1` / ...

**Comment by AlmasB** [comment]
`11+dev-SNAPSHOT` should be available. Please can you check all is fine from your side.

**Comment by sormuras** [comment]
Manually downloaded `fxgl-11+dev-20210729.163945-1.jar` from https://oss.sonatype.org/content/repositories/snapshots/com/github/almasb/fxgl/11+dev-SNAPSHOT/ and ran `jar --describe-module --file fxgl-11+dev-20210729.163945-1.jar` produced:

```
com.almasb.fxgl.all@11+dev-SNAPSHOT jar:file:///.../fxgl-11+dev-20210729.163945-1.jar!/module-info.class
exports com.almasb.fxgl.app
exports com.almasb.fxgl.app.scene
exports com.almasb.fxgl.app.services
exports com.almasb.fxgl.dev
exports com.almasb.fxgl.dev.editor
exports com.almasb.fxgl.dsl
exports com.almasb.fxgl.dsl.components
exports com.almasb.fxgl.dsl.components.view
exports com.almasb.fxgl.dsl.effects
exports com.almasb.fxgl.dsl.handlers
exports com.almasb.fxgl.dsl.views
exports com.almasb.fxgl.gameplay
exports com.almasb.fxgl.multiplayer
exports com.almasb.fxgl.scene3d
requires com.almasb.fxgl.core transitive
requires com.almasb.fxgl.entity transitive
requires com.almasb.fxgl.gameplay transitive
requires com.almasb.fxgl.io transitive
requires com.almasb.fxgl.scene transitive
requires com.fasterxml.jackson.databind
requires com.gluonhq.attach.lifecycle
requires java.base mandated
requires java.desktop
requires javafx.base transitive
requires javafx.controls transitive
requires javafx.fxml transitive
requires javafx.graphics transitive
qualified opens com.almasb.fxgl.app to com.almasb.fxgl.core
qualified opens com.almasb.fxgl.app.services to com.almasb.fxgl.core
qualified opens com.almasb.fxgl.dev to com.almasb.fxgl.core
qualified opens com.almasb.fxgl.dev.profiling to com.almasb.fxgl.core
qualified opens com.almasb.fxgl.dsl to com.almasb.fxgl.core
qualified opens com.almasb.fxgl.dsl.components to com.almasb.fxgl.core
contains com.almasb.fxgl.generated
contains com.almasb.fxgl.scene3d.obj
contains fxglassets
contains fxglassets.languages
contains fxglassets.textures
contains fxglassets.textures.intro
contains fxglassets.textures.particles
contains fxglassets.ui.css
contains fxglassets.ui.fonts
```

✔

Next ... use `11+dev-SNAPSHOT` in project `bach-fxgl`... _in progress_

**Comment by AlmasB** [comment]
this should do it:
```
<dependencies>
    <dependency>
        <groupId>com.github.almasb</groupId>
        <artifactId>fxgl</artifactId>
        <version>11+dev-SNAPSHOT</version>
    </dependency>
</dependencies>

<repositories>
    <repository>
        <id>oss.sonatype.org-snapshot</id>
        <url>https://oss.sonatype.org/content/repositories/snapshots</url>
    </repository>
</repositories>
```

**Comment by AlmasB** [comment]
Or how do you normally get deps into bach?

**Comment by sormuras** [comment]
...it depends. 😉

**Comment by sormuras** [comment]
Running (in project `bach-fxgl`)

`java --module-path .bach\external-modules --list-modules`

now yields 

```text
java.base@16.0.1
java.compiler@16.0.1
java.datatransfer@16.0.1
java.desktop@16.0.1
java.instrument@16.0.1
java.logging@16.0.1
[...]
jdk.xml.dom@16.0.1
jdk.zipfs@16.0.1
com.almasb.fxgl.all@11+dev-SNAPSHOT
com.almasb.fxgl.core@11+dev-SNAPSHOT
com.almasb.fxgl.entity@11+dev-SNAPSHOT
com.almasb.fxgl.gameplay@11+dev-SNAPSHOT
com.almasb.fxgl.io@11+dev-SNAPSHOT
com.almasb.fxgl.scene@11+dev-SNAPSHOT
com.fasterxml.jackson.annotation@2.12.3
com.fasterxml.jackson.core@2.12.3
com.fasterxml.jackson.databind@2.12.3
com.gluonhq.attach.audio@4.0.11
com.gluonhq.attach.lifecycle@4.0.11
com.gluonhq.attach.storage@4.0.11
com.gluonhq.attach.util@4.0.11
javafx.base
javafx.controls
javafx.fxml
javafx.graphics
javafx.media
kotlin.stdlib
```

✔


Now, enhance JavaFX's and Kotlin's builds to include Java module version information.

**Comment by sormuras** [comment]
> Or how do you normally get deps into bach?

There used to be support for Maven SNAPSHOT version in Bach 1.x ... or was it 2.x?! Today, I (re)solved it via

```java
    Bach.build(
        project ->
            project
                .withName("bach-fxgl")
                .withVersion("1-ea")
                .withExternalModuleLocators(JavaFX.of("16"))
                .withExternalModuleLocators(
                    new ExternalModuleLocation(
                        "com.almasb.fxgl.all",
                        "https://oss.sonatype.org/content/repositories/snapshots/com/github/almasb/fxgl/11+dev-SNAPSHOT/fxgl-11+dev-20210729.163945-1.jar"),
                    new ExternalModuleLocation(
                        "com.almasb.fxgl.core",
                        "https://oss.sonatype.org/content/repositories/snapshots/com/github/almasb/fxgl-core/11+dev-SNAPSHOT/fxgl-core-11+dev-20210729.163947-1.jar"),
                    new ExternalModuleLocation(
                        "com.almasb.fxgl.entity",
                        "https://oss.sonatype.org/content/repositories/snapshots/com/github/almasb/fxgl-entity/11+dev-SNAPSHOT/fxgl-entity-11+dev-20210729.163945-1.jar"),
                    new ExternalModuleLocation(
                        "com.almasb.fxgl.gameplay",
                        "https://oss.sonatype.org/content/repositories/snapshots/com/github/almasb/fxgl-gameplay/11+dev-SNAPSHOT/fxgl-gameplay-11+dev-20210729.163949-1.jar"),
                    new ExternalModuleLocation(
                        "com.almasb.fxgl.io",
                        "https://oss.sonatype.org/content/repositories/snapshots/com/github/almasb/fxgl-io/11+dev-SNAPSHOT/fxgl-io-11+dev-20210729.163943-1.jar"),
                    new ExternalModuleLocation(
                        "com.almasb.fxgl.scene",
                        "https://oss.sonatype.org/content/repositories/snapshots/com/github/almasb/fxgl-scene/11+dev-SNAPSHOT/fxgl-scene-11+dev-20210729.163944-1.jar")
                )
                .withExternalModuleLocators(new SormurasModulesLocator("2021.05.20"))
                .withMainSpace(
                    main ->
                        main.withModule("com.github.sormuras.bach.fxgl")
                            .withModulePaths(".bach/external-modules")),
        args);
```

Which is **not** the final syntax, yet.

**Comment by AlmasB** [comment]
Ok, I think we can call this a success for module version info, thanks again.

**Comment by sormuras** [comment]
Once you released FXGL `11.8` to Maven Central, I'll be able to use `FXGL.of("11.8")` again.
